### ISSUE TITLE
When is amplify going to support the new js resolvers 

### ISSUE THREAD

**Issue by austinamorusoyardstick** [aws-amplify/amplify-category-api#1015]

**Comment by isaac-elvt** [comment]
bump

**Comment by BBopanna** [comment]
Need this soon please!

**Comment by james-kuihi** [comment]
Third'd - this would be a great feature to have!

**Comment by danfreid** [comment]
Would be much appreciated by developers

**Comment by alexboulay** [comment]
We would appreciate this too!

**Comment by vaughngit** [comment]
bump!

**Comment by tminus1-design** [comment]
bump

**Comment by prashant-joshi-25** [comment]
bump!!

**Comment by mithun35h** [comment]
bump

**Comment by austinamorusoyardstick** [comment]
Honestly changing all the vlt templates to JS would likely enable to Amplify community to issue patches easier and fixes / plugins for issues surrounding resolvers.

**Comment by biller-aivy** [comment]
Any plans for this rebuild of the resolvers?

**Comment by jay-herrera** [comment]
bump

**Comment by barshopen** [comment]
Seems like AWS phases out of VTL if favor of APPSYNC_JS (source: https://docs.aws.amazon.com/appsync/latest/devguide/resolver-mapping-template-reference.html),
<img width="1447" alt="image" src="https://github.com/aws-amplify/amplify-category-api/assets/44550986/e7c6682c-e85a-4b91-ba89-a81091329052">
are there any updates on this feature?

**Comment by chrisl777** [comment]
bump

**Comment by giraudvalentin** [comment]
its a nightmare to do graphql with appsync.. i update to node 16, update my sls to 3, update appsync plugin to support js resolver and, what, the simulator broke everything

**Comment by renebrandel** [comment]
Hi - while we update our documentation to illustrate how to do this with the Amplify CLI, I do want to bring attention to the fact that the GraphQL API category is now available first-class as a CDK construct. In the CDK construct, you can use JavaScript resolvers. Check out our announcement post here: https://aws.amazon.com/blogs/mobile/announcing-aws-amplifys-graphql-api-cdk-construct-deploy-real-time-graphql-api-and-data-stack-on-aws/

**Comment by ElliottAtYardstick** [comment]
Yes please

**Comment by travishaby** [comment]
This would be lovely, yes please!

**Comment by lukasulbing** [comment]
bump

**Comment by Dennis-Dekker** [comment]
bump

**Comment by espetro** [comment]
bump 🤭

**Comment by renebrandel** [comment]
Hi - @armenr - sorry for the dropping the ball here. So here's a working sample that should work for you. For the JS resolvers. I build a quick "echo this message" resolver. The graphQL schema looks like this:

```
input AMPLIFY { globalAuthRule: AuthRule = { allow: public } } # FOR TESTING ONLY!

type Todo @model {
  id: ID!
  name: String!
  description: String
}

type Query {
  echo(message: String): String # your custom queries here
}
```

The custom stack looks like this:
```ts
import * as cdk from 'aws-cdk-lib';
import * as AmplifyHelpers from '@aws-amplify/cli-extensibility-helper';
import * as appsync from 'aws-cdk-lib/aws-appsync';
import { AmplifyDependentResourcesAttributes } from '../../types/amplify-dependent-resources-ref';
import { Construct } from 'constructs';

const jsResolverTemplate = `
export function request(ctx) {
  return {
    payload: null
  }
}

export function response(ctx) {
  return ctx.arguments.message
}
`

export class cdkStack extends cdk.Stack {
  constructor(
    scope: Construct,
    id: string,
    props?: cdk.StackProps,
    amplifyResourceProps?: AmplifyHelpers.AmplifyResourceProps
  ) {
    super(scope, id, props);
    /* Do not remove - Amplify CLI automatically injects the current deployment environment in this input parameter */
    new cdk.CfnParameter(this, 'env', {
      type: 'String',
      description: 'Current Amplify CLI env name'
    });

    // Access other Amplify Resources
    const retVal: AmplifyDependentResourcesAttributes =
      AmplifyHelpers.addResourceDependency(
        this,
        amplifyResourceProps.category,
        amplifyResourceProps.resourceName,
        [
          {
            category: 'api',
            resourceName: 'gen1jsresolver'
          }
        ]
      );

    const resolver = new appsync.CfnResolver(this, 'CustomResolver', {
      // apiId: retVal.api.new.GraphQLAPIIdOutput,
      // https://github.com/aws-amplify/amplify-cli/issues/9391#event-5843293887
      // If you use Amplify you can access the parameter via Ref since it's a CDK parameter passed from the root stack.
      // Previously the ApiId is the variable Name which is wrong , it should be variable value as below
      apiId: cdk.Fn.ref(retVal.api.gen1jsresolver.GraphQLAPIIdOutput),
      fieldName: 'echo',
      typeName: 'Query', // Query | Mutation | Subscription
      code: jsResolverTemplate,
      dataSourceName: 'NONE_DS', // DataSource name
      runtime: {
        name: 'APPSYNC_JS',
        runtimeVersion: '1.0.0'
      }
    });
  }
}
```


Do you mind sharing more about the exact error you're facing? My guess is that this line is probably where the error is coming from:
```
      graphqlApiId: apiResourceReference.api.myprojectV2AwesomeAPI.GraphQLAPIIdOutput,
```

My guess is that the resolution of the reference doesn't happen correctly, thus the API won't be referenced. It should be this:

```
      graphqlApiId: cdk.Fn.ref(apiResourceReference.api.myprojectV2AwesomeAPI.GraphQLAPIIdOutput),
```

**Comment by renebrandel** [comment]
@armenr - also cut a docs PR https://github.com/aws-amplify/docs/pull/7087

**Comment by renebrandel** [comment]
That's correct. We're making this experience significantly more seamless in Gen 2. That should help in the future. Gen 2 is still in dev preview.  https://docs.amplify.aws/gen2/build-a-backend/data/custom-business-logic/

**Comment by renebrandel** [comment]
Hi - let's open a separate issue on the amplify-backend repo to discuss Gen 2 issues if this response isn't sufficient for now. Just to keep this thread focused around JS resolvers.

We're working on the migration strategy for Gen 1 to Gen 2. At the time of Gen 2 GA, you'll be able to rebuild most use cases with Gen 2 and build out significantly more (for example, JS resolvers). We're building up detailed feature matrix, so customers will understand which use case is supported in:
- in both Gen 1 and Gen 2
- only in Gen 2
- only in Gen 1 - in this case, we'll document workarounds or rationale as why the use case isn't valid in Gen 2

After Gen 2 GA, our highest priority item is building a migration path for auth, data, and storage. Specifically, our plan is to migrate the stateful resources from your Gen 1 project to Gen 2. I.e. porting over your user pool, database tables, buckets. For "stateless" services (Lambda functions, AppSync API), we'll provide a guide on how to recreate in functional parity in Gen 2.

**Comment by Dennis-Dekker** [comment]
Thank you both for the input on this thread, I have just caught up with the issue after my holiday.

We are currently also working on getting the JS resolvers up and running and were also able to get this working for queries and mutations that we manually added to the schema. However, we would like to replace the amplify generated VTL resolvers that are made during the build. Do you have any guidance on how to do this? I created an issue for this some time ago but haven't received any replies on that: #2211

**Comment by naedx** [comment]
I'm also interested in this use case. Typescript is integral to how we work with JS code for Lambda, etc.

**Comment by renebrandel** [comment]
@armenr - sorry for the late reply. The core difference is that the `CfnResolver` is a "L1 construct" vs. `Resolver` is an "L2 construct". Using the L1 construct in Gen 1 is simpler for referencing the API but you're not prohibited to use the L2 construct. Either case though the TypeScript compilation workflow would likely remain the same using your suggestion of leveraging a `prePush` command hook.
